imap_processing/lo/l1b/lo_l1b.py

```diff
@@ -368,6 +382,195 @@ def set_spin_cycle(
     return l1b_de
 
 
+# TODO: The spin cycle function above needs to be updated for DEs. We cannot assume
+#  there are 28 spins per ASC and we should calculate the spin start number based on the
+#  corresponding L1A spin data Acq Start for the ASC. The implementation below should be
+#  should be used for the DE rather than the above function, but in the interest of time
+#  the below function is only hooked up to the histogram rates processing and should be
+#  integrated into the DE processing in a later PR.
+# TODO: Break up the invalid spin ASC removal and the code to find the closest DE/Hist
+#  and spin ASCs into their own functions.
+def set_spin_cycle_from_spin_data(
+    l1a_science: xr.Dataset, l1b_science: xr.Dataset, spin_data: xr.Dataset
+) -> xr.Dataset:
+    """
+    Set the spin cycle for each direct event using the L1A spin data.
+
+    The spin cycle is the average spin for a given Aggregated Science Cycle
+     in a given ESA Step.
+
+    Parameters
+    ----------
+    l1a_science : xr.Dataset
+        The L1A Histogram or Direct Event dataset.
+    l1b_science : xr.Dataset
+        The L1B Histogram Rate or Direct Event dataset.
+    spin_data : xr.Dataset
+        The L1A Spin dataset.
+
+    Returns
+    -------
+    l1b_science : xr.Dataset
+        The L1B science dataset with the spin cycle added for each direct event.
+    """
+    acq_start, _acq_end = convert_start_end_acq_times(spin_data)
+
+    spin_met_per_asc = spin_data["shcoarse"].values.astype(np.float64)
+    science_met_per_asc = ttj2000ns_to_met(l1a_science["epoch"]).astype(np.float64)
+
+    closest_start_acq_indices, time_diff = match_science_to_spin_asc(
```

Comment:
I don't think this is actually the start_acq indices, because you are matching on the SHCOARSE values which were timestamped after the full acquisition of each of the packets right?
```suggestion
    science_to_spin_indices, time_diff = match_science_to_spin_asc(
```

Reply:
yeah, you're right, I'll update that name